### PR TITLE
Update optimizer to AdamW for rtmdet_inst_tiny

### PR DESCRIPTION
**Pull request by AlbertvanHouten**
### Summary
The RTMdet recipe was implemented with the SGD optimizer, however, this does not match with the default setup described in the [paper](https://ar5iv.labs.arxiv.org/html/2212.07784) where they use AdamW. 

I have ran a few benchmarks using the WGISD dataset and the map50 improved significantly when using AdamW. The bottom result in the screenshot is with SGD optimizer, other results are using AdamW. 
image

### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [ ] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml [hunk 1]
```diff
@@ -6,11 +6,11 @@ model:
     model_name: rtmdet_inst_tiny
 
     optimizer:
-      class_path: torch.optim.SGD
+      class_path: torch.optim.AdamW
       init_args:
-        lr: 0.001
-        momentum: 0.9
-        weight_decay: 0.0001
+        lr: 0.004
```
Comment (kprokofi) [inline library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml]: Is this value for learning rate was chosen based only on 1 dataset? Did you check that this value improves metric consistently for at least a couple of datasets?

**Review by kprokofi — COMMENTED** [review]
Thanks for noticing that problem with AdamW

File: library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml (hunk 1 above)
Comment (AlbertvanHouten) [reply library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml]: This value comes from the paper linked in the PR description

Comment (leoll2) [inline library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml]: Please also update the values in the model manifest: 

https://github.com/open-edge-platform/training_extensions/blob/2e32abb6346b4061e42850d530a11c49426134a3/application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml#L83:L84

File: application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml [hunk 2]
```diff
@@ -81,8 +81,8 @@ hyperparameters:
         sharpness: 0.5
         probability: 0.5
   training:
-    learning_rate: 0.001
-    weight_decay: 0.0001
+    learning_rate: 0.004
+    weight_decay: 0.01
```
Comment (leoll2) [inline application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: I tested the new learning rate and weight decay values on a demo dataset, and got worse result; both runs use AdamW. Of course it's only one experiment and not statistically relevant, but we should be careful to modify these values. I'll repeat the experiment on other datasets too.

**Old hyperparameter values**

Image

Image

**New hyperparameter values**

Image

Image

Comment (AlbertvanHouten) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: The old learning rate value is more favourable for smaller datasets while the new value is better for larger datasets. This will always be a trade-off. Ideally we have some sort of mechanism that adjusts the starting learning rate based on the dataset size. For now, I have copied the values defined in the paper (I did reduce weight decay slightly to be inline with other model recipes).

Comment (AlbertvanHouten) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: You can also see the learning rate is still incrementing meaning it is still warming up. It should never finish training before warmup has finished so that is an additional bug. I don't think we adjust warmup steps based on dataset size either. @kprokofi

Comment (leoll2) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: > For now, I have copied the values defined in the paper

Research papers tend to optimize for their benchmark datasets, usually large datasets like COCO. Most Geti users use datasets with 50-300 images (observed in Geti Cloud), so we should tune the hyperparameters based on medium-sized datasets, imho.

> You can also see the learning rate is still incrementing meaning it is still warming up. It should never finish training before warmup has finished so that is an additional bug

I agree, it's a known problem that some models have suboptimal LR schedules. @kprokofi please address it in scope of https://github.com/open-edge-platform/training_extensions/issues/5072

Comment (leoll2) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: I've just tested on another dataset with 200 images. Again, the old LR results in a better mAP and healthier learning curves. In this case, a too aggressive LR caused the training loss to quickly plateau at a relatively high level, and the training terminated prematurely because of early stopping.

**Old values**

image

**New values**

image

Comment (AlbertvanHouten) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: I have reduced the learning rate to 0.0007. The paper uses 0.004 for batch size 256 but we are using batch size 8. It seems common to use square root scaling for AdamW so for batch size of 8 we get  0.004 * sqrt(8/256) ~= 0.0007. I will run another check with the WGISD dataset.

File: library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml (hunk 1 above)
Comment (kprokofi) [reply library/src/getitune/recipe/instance_segmentation/rtmdet_inst_tiny.yaml]: The values in papers are always optimal only for COCO dataset. It is sad, but all academic setups just want to show the best metric on COCO without real application. Anyway, I think you already discussed that with Leo below.

File: application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml (hunk 2 above)
Comment (AlbertvanHouten) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: Ran 2 tests, both got slightly better results with new learning rate:

image

Comment (leoll2) [reply application/backend/app/supported_models/manifests/instance_segmentation/rtmdet_tiny.yaml:85]: I also get better results with the the latest LR values:

image

And significantly better than with SGD.

**Review by leoll2 — APPROVED** [review]
